codelabs/Utopia Wholesale Codelab/Advanced Features/Deploying to Google Cloud Run.md

```diff
@@ -0,0 +1,279 @@
+* How to deploy servers to Google Cloud Run
+* How to connect the OpenID4VCI server to the Records server
+
+---
+
+## **What You'll Need**
+
+* Access to the `identity-credential` repository
+* Google Cloud project with Cloud Run API enabled
+* Terminal/command line access
+* Java 17+ (for building the JAR files)
+
+---
+
+## **Configuration Changes**
+
+In [this commit](https://github.com/hanluOMH/identity-credential/commit/bd51ad7dc8fd128491da7dff709e7f36eda30330) from the [`deploy_cloud` branch](https://github.com/hanluOMH/identity-credential/tree/deploy_cloud), several configuration changes were made to prepare the servers for Cloud Run deployment:
+
+### **Changes to `multipaz-openid4vci-server`**
+
+The following changes were made to `multipaz-openid4vci-server/src/main/resources/resources/default_configuration.json`:
+
+1. **Port Configuration**: Changed `server_port` from `8007` to `8080`
+   - Cloud Run expects services to listen on port 8080 by default
+   - The PORT environment variable can override this if needed
+
+2. **Base URL Configuration**: Added `base_url` pointing to the Cloud Run deployment URL
+   ```json
+   "base_url": "https://multipaz-openid4vci-server-971523157550.us-central1.run.app"
+   ```
+   - **Important**: Replace this with your own Cloud Run deployment URL after deploying the server
+   - This ensures all redirects and generated URLs use the correct public URL
+   - Without this, the server would generate URLs with `localhost:8080`
+
+3. **System of Record URL**: Added `system_of_record_url` pointing to the Records server
+   ```json
+   "system_of_record_url": "https://multipaz-records-server-971523157550.us-central1.run.app"
+   ```
+   - **Important**: Replace this with your own Records server Cloud Run URL after deploying it
+   - This connects the OpenID4VCI server to the Records server
+   - The OpenID4VCI server uses this to retrieve identity data when issuing credentials
+
+### **Changes to `multipaz-records-server`**
+
+The following changes were made to `multipaz-records-server/src/main/resources/resources/default_configuration.json`:
+
+1. **Port Configuration**: Changed `server_port` from `8004` to `8080`
+   - Consistent with Cloud Run's expected port
+
+2. **Base URL Configuration**: Added `base_url` pointing to the Cloud Run deployment URL
+   ```json
+   "base_url": "https://multipaz-records-server-971523157550.us-central1.run.app"
+   ```
+   - **Important**: Replace this with your own Cloud Run deployment URL after deploying the server
+   - Ensures proper redirects when accessing the root URL
+
+3. **Admin Password**: Added fixed `admin_password` set to `"multipaz-records"`
+   - Previously, the server generated a random password on each startup
+   - This makes the password predictable and manageable for deployment
+
+---
+
+## **Deployment Steps**
+
+### **Step 1: Create Dockerfile**
+
+Create a Dockerfile for the OpenID4VCI server in the `multipaz-openid4vci-server` directory:
+
+```bash
```

Comment:
for some reason the highlighting doesn't work - can you also try using `shell` instead of `bash`?